### PR TITLE
HTTP Semconv migration Part3 Server - v1.24.0 support

### PR DESCRIPTION
**Pull request by MadVikingGod**
This change adds the new semantic version (v1.24.0) attribute producer to the semconv of otlehttp.

The full PR is https://github.com/open-telemetry/opentelemetry-go-contrib/pull/5092
Part of https://github.com/open-telemetry/opentelemetry-go-contrib/issues/5331

### CODE REVIEW

**Comment by MadVikingGod** [comment]
Changelog will be part of follow on PR as part of #5331

**Review by dashpole — APPROVED** [review]
Could use a few more test cases, but LGTM otherwise

File: instrumentation/net/http/otelhttp/internal/semconv/v1.24.0.go [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+// nginx server. More generically, the primary server name would be the host
+// header value that matches the default virtual host of an HTTP server. It
+// should include the host identifier and if a port is used to route to the
+// server that port identifier should be included as an appropriate port
+// suffix.
+//
+// If the primary server name is not known, server should be an empty string.
+// The req Host will be used to determine the server instead.
+func (n newHTTPServer) RequestTraceAttrs(server string, req *http.Request) []attribute.KeyValue {
+	const MaxAttributes = 11
```
Comment (MrAlias) [inline instrumentation/net/http/otelhttp/internal/semconv/v1.24.0.go:36]:
```suggestion
	const maxAttributes = 11
```

Comment (MrAlias) [reply instrumentation/net/http/otelhttp/internal/semconv/v1.24.0.go:36]: This is likely going to over allocate. Can we pre-determine the correct size or have an issue to track doing so later?

File: instrumentation/net/http/otelhttp/internal/semconv/v1.24.0.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+	attrs := make([]attribute.KeyValue, MaxAttributes)
+	var host string
+	var p int
+	if server == "" {
+		host, p = splitHostPort(req.Host)
+	} else {
+		// Prioritize the primary server name.
+		host, p = splitHostPort(server)
+		if p < 0 {
```
Comment (MrAlias) [inline instrumentation/net/http/otelhttp/internal/semconv/v1.24.0.go:45]:
```suggestion
		if host == "" && p < 0 {
```

File: instrumentation/net/http/otelhttp/internal/semconv/v1.24.0.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+// header value that matches the default virtual host of an HTTP server. It
+// should include the host identifier and if a port is used to route to the
+// server that port identifier should be included as an appropriate port
+// suffix.
+//
+// If the primary server name is not known, server should be an empty string.
+// The req Host will be used to determine the server instead.
+func (n newHTTPServer) RequestTraceAttrs(server string, req *http.Request) []attribute.KeyValue {
+	const MaxAttributes = 11
+	attrs := make([]attribute.KeyValue, MaxAttributes)
```
Comment (MrAlias) [inline instrumentation/net/http/otelhttp/internal/semconv/v1.24.0.go:37]:
```suggestion
	attrs := make([]attribute.KeyValue, 0, MaxAttributes)
```

Tracking append position could then be replaced with `append`

File: instrumentation/net/http/otelhttp/internal/semconv/v1.24.0.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+	}
+
+	if attr, ok := methodLookup[strings.ToUpper(method)]; ok {
+		attrs[0] = attr
+	} else {
+		// If the Original methos is not a standard HTTP method fallback to GET
+		attrs[0] = semconvNew.HTTPRequestMethodGet
+	}
+	attrs[1] = semconvNew.HTTPRequestMethodOriginal(method)
+	return 2
```
Comment (MrAlias) [inline instrumentation/net/http/otelhttp/internal/semconv/v1.24.0.go:116]: The index state is not tracked with a variable. This is going to be brittle and tough to evolve.

File: instrumentation/net/http/otelhttp/internal/semconv/v1.24.0.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+		attrs[i] = semconvNew.NetworkProtocolVersion(protoVersion)
+		i++
+	}
+
+	return slices.Clip(attrs[:i])
+}
+
+func (n newHTTPServer) method(method string, attrs []attribute.KeyValue) int {
+	if method == "" {
+		attrs[0] = semconvNew.HTTPRequestMethodGet
```
Comment (MrAlias) [inline instrumentation/net/http/otelhttp/internal/semconv/v1.24.0.go:101]: It is not obvious that `attrs` passed here has to be pre-allocated.

Using an `append` approach would avoid the panic if `len(attrs) <= 0`.

File: instrumentation/net/http/otelhttp/internal/semconv/v1.24.0.go [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+	}
+	attrs[0] = semconvNew.URLScheme("http")
+	return 1
+}
+
+// TraceResponse returns trace attributes for telemetry from an HTTP response.
+//
+// If any of the fields in the ResponseTelemetry are not set the attribute will be omitted.
+func (n newHTTPServer) ResponseTraceAttrs(resp ResponseTelemetry) []attribute.KeyValue {
+	attributes := []attribute.KeyValue{}
```
Comment (MrAlias) [inline instrumentation/net/http/otelhttp/internal/semconv/v1.24.0.go:132]: The allocation strategy is not the same as `RequestTraceAttrs`.

**Review by MrAlias — COMMENTED** [review]
There is missing test coverage.

File: instrumentation/net/http/otelhttp/internal/semconv/v1.24.0.go [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+	}
+	if attr, ok := methodLookup[method]; ok {
+		attrs[0] = attr
+		return 1
+	}
+
+	if attr, ok := methodLookup[strings.ToUpper(method)]; ok {
+		attrs[0] = attr
+	} else {
+		// If the Original methos is not a standard HTTP method fallback to GET
```
Comment (kevinburkesegment) [inline instrumentation/net/http/otelhttp/internal/semconv/v1.24.0.go:112]:
```suggestion
		// If the Original method is not a standard HTTP method fallback to GET
```

File: instrumentation/net/http/otelhttp/internal/semconv/v1.24.0_test.go [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+				attribute.Int("http.response.body.size", 802),
+				attribute.Int("http.response.status_code", 200),
+			},
+		},
+		{
+			name: "with errors",
+			resp: ResponseTelemetry{
+				StatusCode: 200,
+				ReadBytes:  701,
+				ReadError:  fmt.Errorf("read error"),
```
Comment (kevinburkesegment) [inline instrumentation/net/http/otelhttp/internal/semconv/v1.24.0_test.go:66]:
```suggestion
				ReadError:  errors.New("read error"),
```

File: instrumentation/net/http/otelhttp/internal/semconv/v1.24.0.go (hunk 2 above)
Comment (MadVikingGod) [reply instrumentation/net/http/otelhttp/internal/semconv/v1.24.0.go:42]: The goal of this if branch is only to change the port if one wasn't part of `server`.  This was logic from semconvutil: https://github.com/open-telemetry/opentelemetry-go-contrib/blob/261b731dff1e55aaa151086bcb8aee676359db61/instrumentation/net/http/otelhttp/internal/semconvutil/httpconv.go#L334-L344

File: instrumentation/net/http/otelhttp/internal/semconv/v1.24.0.go (hunk 3 above)
Comment (MadVikingGod) [reply instrumentation/net/http/otelhttp/internal/semconv/v1.24.0.go]: This was replaced with logic similar to what is used in semconvutil.  Please re-review

File: instrumentation/net/http/otelhttp/internal/semconv/v1.24.0.go (hunk 6 above)
Comment (MadVikingGod) [reply instrumentation/net/http/otelhttp/internal/semconv/v1.24.0.go]: This was replaced with logic similar to what is used in semconvutil.  Please re-review

File: instrumentation/net/http/otelhttp/internal/semconv/v1.24.0.go (hunk 4 above)
Comment (MadVikingGod) [reply instrumentation/net/http/otelhttp/internal/semconv/v1.24.0.go]: This was replaced with logic similar to what is used in semconvutil.  Please re-review

**Review by XSAM — APPROVED** [review]

File: instrumentation/net/http/otelhttp/internal/semconv/v1.24.0_test.go (hunk 8 above)
Comment (MadVikingGod) [reply instrumentation/net/http/otelhttp/internal/semconv/v1.24.0_test.go:75]: I will keep this so it has parity with the tests in v1.20.0.

**Pull request merged**